File: cpp/algorithms/dynamic_programming/knapsack/knapsack_dp.cpp
#include <bits/stdc++.h>
using namespace std;

int knapsack01(const vector<int>& w, const vector<int>& v, int cap) {
    vector<int> dp(cap + 1, 0);
    for (int i = 0; i < (int)w.size(); ++i)
        for (int x = cap; x >= w[i]; --x) dp[x] = max(dp[x], dp[x - w[i]] + v[i]);
    return dp[cap];
}

int knapsack_unbounded(const vector<int>& w, const vector<int>& v, int cap) {
    vector<int> dp(cap + 1, 0);
    for (int x = 1; x <= cap; ++x)
        for (int i = 0; i < (int)w.size(); ++i)
            if (x >= w[i]) dp[x] = max(dp[x], dp[x - w[i]] + v[i]);
    return dp[cap];
}

int main() {
    vector<int> w{1, 2, 3}, val{6, 10, 12};
    assert(knapsack01(w, val, 5) == 22);
    assert(knapsack_unbounded({1, 3, 4}, {15, 50, 60}, 8) == 130);
    cout << "knapsack_dp OK" << endl;
    return 0;
}

File: python/algorithms/dynamic_programming/knapsack/knapsack_dp.py
"""0-1 背包与完全背包。"""

from __future__ import annotations


def knapsack_01(weights: list[int], values: list[int], cap: int) -> int:
    dp = [0] * (cap + 1)
    for w, v in zip(weights, values):
        for x in range(cap, w - 1, -1):
            dp[x] = max(dp[x], dp[x - w] + v)
    return dp[cap]


def knapsack_unbounded(weights: list[int], values: list[int], cap: int) -> int:
    dp = [0] * (cap + 1)
    for x in range(1, cap + 1):
        for w, v in zip(weights, values):
            if x >= w:
                dp[x] = max(dp[x], dp[x - w] + v)
    return dp[cap]


if __name__ == "__main__":
    w = [1, 2, 3]
    val = [6, 10, 12]
    assert knapsack_01(w, val, 5) == 22
    assert knapsack_unbounded([1, 3, 4], [15, 50, 60], 8) == 130
    assert knapsack_01([], [], 0) == 0
    print("knapsack_dp OK")
